cargo/cargo_build_script.bzl

```diff
@@ -0,0 +1,108 @@
+load("@io_bazel_rules_rust//rust:private/rustc.bzl", "BuildInfo", "rustc_compile_action")
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+def _build_script_impl(ctx):
+    toolchain = find_toolchain(ctx)
+    out_dir = ctx.actions.declare_directory(ctx.label.name + ".out_dir")
+    env_out = ctx.actions.declare_file(ctx.label.name + ".env")
+    flags_out = ctx.actions.declare_file(ctx.label.name + ".flags")
+    env = {
+        "CARGO_MANIFEST_DIR": ctx.build_file_path.rsplit("/", 1)[0],
+        "RUSTC": toolchain.rustc.path,
+        "TARGET": toolchain.target_triple,
+        "OUT_DIR": out_dir.path,
+        "BINARY_PATH": ctx.executable.script.path,
+    }
+
+    for f in ctx.attr.crate_features:
+        env["CARGO_FEATURE_" + f.upper().replace("-", "_")] = "1"
+
+    ctx.actions.run(
+        executable = ctx.executable._cargo_build_script_runner,
+        arguments = [ctx.executable.script.path, env_out.path, flags_out.path],
+        outputs = [out_dir, env_out, flags_out],
+        tools = [ctx.executable.script, ctx.executable._cargo_build_script_runner],
+        inputs = ctx.files.srcs + [ctx.executable.script, toolchain.rustc],
+        mnemonic = "CargoBuildScriptRun",
+        env = env,
+    )
+
+    return [
+        BuildInfo(
+            out_dir = out_dir,
+            rustc_env = env_out,
+            flags = flags_out,
+        ),
+    ]
+
+cargo_build_script_run = rule(
+    _build_script_impl,
+    attrs = {
+        "script": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            cfg = "host",
+            doc = "The binary script to run, generally a rust_binary target. ",
+        ),
+        "crate_features": attr.string_list(doc = "The list of rust features that the build script should consider activated."),
+        "srcs": attr.label_list(allow_files = True, doc = "A list of files the scripts needs to access, generally the sources of the crate for which the build script is working"),
+        "_cargo_build_script_runner": attr.label(
+            executable = True,
+            allow_files = True,
+            default = Label("//cargo/cargo_build_script_runner:cargo_build_script_runner"),
+            cfg = "host",
+        ),
+    },
+    toolchains = [
+        "@io_bazel_rules_rust//rust:toolchain",
+    ],
+    doc = """
+Execute a rust build script to generate build attributes.
```

Comment:
Thanks for the docs; this helps clarify things a lot.

Do you have a corresponding cargo raze patch already? If not, I'm very happy to start making templates that generate BUILD files in this format.

Reply:
No I have not since I have created those orginally trying to get rid totally of cargo raze.